File: backend/src/core/operation.ts
import assert from "assert";
import { stringify } from "canonical-json";
import { z, ZodType } from "zod";
import Tree from "./tree.js";

export enum OperationType {
    ADD_NODE = "add_node",
    REOPEN_NODE = "reopen_node",
    COMPLETED_NODE = "complete_node",
    REMOVE_NODE = "remove_node",
    REMOVE_SUBTREE = "remove_subtree",
    MOVE_NODE = "move_node",
}

enum PseudoOperationType {
    UNDO = "undo",
    FLUSH = "flush",
}

export enum ExtOperationType {
    ADD_NODE = OperationType.ADD_NODE,
    REOPEN_NODE = OperationType.REOPEN_NODE,
    COMPLETED_NODE = OperationType.COMPLETED_NODE,
    REMOVE_NODE = OperationType.REMOVE_NODE,
    REMOVE_SUBTREE = OperationType.REMOVE_SUBTREE,
    MOVE_NODE = OperationType.MOVE_NODE,

    // extended
    UNDO = PseudoOperationType.UNDO,
    FLUSH = PseudoOperationType.FLUSH,
}

const methodNameMap: Record<OperationType, keyof Tree> = {
    [OperationType.ADD_NODE]: "addNode",
    [OperationType.REOPEN_NODE]: "reopenNode",
    [OperationType.COMPLETED_NODE]: "completeNode",
    [OperationType.REMOVE_NODE]: "removeNode",
    [OperationType.REMOVE_SUBTREE]: "removeSubtree",
    [OperationType.MOVE_NODE]: "moveNode",
}

const methodParamsMap: Record<OperationType, ZodType> = {
    [OperationType.ADD_NODE]: z.object({
        parentNodeId: z.string(),
        newNodeName: z.string(),
        newNodeId: z.string().optional(),
    }),

    [OperationType.REOPEN_NODE]: z.object({
        nodeId: z.string(),
    }),

    [OperationType.COMPLETED_NODE]: z.object({
        nodeId: z.string(),
    }),

    [OperationType.REMOVE_NODE]: z.object({
        nodeId: z.string(),
    }),

    [OperationType.REMOVE_SUBTREE]: z.object({
        nodeId: z.string(),
    }),

    [OperationType.MOVE_NODE]: z.object({
        nodeId: z.string(),
        newParentId: z.string(),
    })
}

interface OperationAttrs {
    opType: OperationType,
    payload: object,
    timestamp: number,
}

export default class Operation {
    opType: OperationType;
    payload: object;
    timestamp: number;

    constructor({
        opType,
        payload,
        timestamp,
    }: OperationAttrs) {
        this.opType = opType;
        this.payload = payload;
        this.timestamp = timestamp;
    }

    stringify(): string {
        return stringify({
            op_type: this.opType,
            payload: this.payload,
            timestamp: this.timestamp,
        });
    }

    apply(tree: Tree): number {
        let method = tree[methodNameMap[this.opType]];
        assert(typeof method === "function");

        const schema = methodParamsMap[this.opType];
        const payload = schema.parse(this.payload);

        // @ts-expect-error: this is already safe, cuz we've checked it with zod schema
        return method.call(tree, payload);
    }
}

File: desktop/app/history/core/operation.py
from enum import Enum
from pydantic import TypeAdapter, ValidationError
from typing import TypedDict, Union, Any, Optional
from .tree import Tree
import json

class OperationType(Enum):
    ADD_NODE = "add_node"
    REOPEN_NODE = "reopen_node"
    COMPLETE_NODE = "complete_node"
    REMOVE_NODE = "remove_node"
    REMOVE_SUBTREE = "remove_subtree"
    MOVE_NODE = "move_node"


class AddNodePayload(TypedDict):
    parent_node_id: str
    new_node_name: str
    new_node_id: str
class ReopenNodePayload(TypedDict):
    node_id: str
class CompleteNodePayload(TypedDict):
    node_id: str
class RemoveNodePayload(TypedDict):
    node_id: str
class RemoveSubtreePayload(TypedDict):
    node_id: str
class MoveNodePayload(TypedDict):
    node_id: str
    new_parent_id: str

OperationPayload = {
    "add_node": AddNodePayload,
    "reopen_node": ReopenNodePayload,
    "complete_node": CompleteNodePayload,
    "remove_node": RemoveNodePayload,
    "remove_subtree": RemoveSubtreePayload,
    "move_node": MoveNodePayload,
}
OperationPayloadUnion = Union[AddNodePayload,
                ReopenNodePayload,
                CompleteNodePayload,
                RemoveNodePayload,
                RemoveSubtreePayload,
                MoveNodePayload]

class Operation:
    """
    Operation is one of the core classes in this application.
    Operation declares how a Tree behaves, and comprises all the history.
    Operations are stored in pending queue and confirmed history.
    """
    def __init__(self,
                 op_type: OperationType,
                 payload: OperationPayloadUnion,
                 timestamp: int):
        self.op_type = op_type
        self.payload = payload
        self.timestamp = timestamp
    
    def to_dict(self) -> dict[str, Any]:
        return {
            "op_type": self.op_type.value,
            "payload": self.payload,
            "timestamp": self.timestamp,
        }
    
    def stringify(self) -> str:
        return json.dumps(self.to_dict(), sort_keys=True, separators=(',', ':'), ensure_ascii=False)

    def apply(self, tree: Tree):
        method = getattr(tree, self.op_type.value)
        res = method(**self.payload)
        return res

def parse_operation(op_str: str) -> Optional[Operation]:
    try:
        data = json.loads(op_str)
    except json.JSONDecodeError:
        return None
    
    if not ("op_type" in data and "payload" in data and "timestamp" in data \
            and isinstance(data["op_type"], str) and isinstance(data["payload"], dict) and isinstance(data["timestamp"], int)):
        return None

    try:
        op_type = OperationType(data["op_type"])
    except ValueError:
        return None
    
    operation_validator = TypeAdapter(OperationPayload[data["op_type"]])
    try:
        parsed_payload = operation_validator.validate_python(data["payload"])
    except ValidationError as e:
        return None
    
    return Operation(op_type=op_type,
                     payload=data["payload"],
                     timestamp=data["timestamp"])
